### ISSUE TITLE
No support for username/password from October ‘22

### ISSUE THREAD

**Issue by SnorreSelmer** [xperimental/netatmo-exporter#26]
I got this email yesterday:

```
To improve the security of our products, we inform you that the Client Credentials grant type method will be completely removed. It will no longer be possible to authenticate with the username and password of the user. 

The effective date of this update is October 2022.

How can you authenticate with Netatmo API ?
From this date, the OAuth2 authorization code flow must be followed for authentication.
```

Is this exporter ready for this change?
What needs to be altered in the compose-file?

**Comment by xperimental** [comment]
I got this email today as well, as I have not read through the documentation yet, I can't say what the changes will be.

I would assume that this means changing how the exporter does authentication and will result in code and configuration changes. There's still some time, though :slightly_smiling_face:

**Comment by SnorreSelmer** [comment]
Excellent! Good to see you’re aware! I love this exporter.

**Comment by mpitt** [comment]
Any news? October is getting closer… 😅

**Comment by theonlydoo** [comment]
[Netatmo seems to have given up on upgrades](https://github.com/Netatmo), they don't support their own 2FA

**Comment by xperimental** [comment]
> Any news? October is getting closer… sweat_smile

I did realize this last weekend as well and started looking into it. Need to update the `netatmo-api-go` library as well. I think I have an idea on how to approach this in a user-friendly way, but overall the authentication flow is not really suited for server-side software unless you have some central provisioning which does not really match the purpose of this project.

**Comment by mpitt** [comment]
>  overall the authentication flow is not really suited for server-side software

Yes that was my impression too, unfortunately. Thanks for the update and for your work!

**Comment by xperimental** [comment]
I plan to have two options:
- a "companion mode" where you do the login flow on another device and then copy the token into the exporter's configuration
- an "interactive mode" where the exporter provides a web-UI to do the normal interactive flow

There are "ideas" right now, nothing implemented yet though, so things might change. I also do not know how long the authentication will be valid and how long the exporter would be able to extend/refresh the token. The exporter might also be able to store the token persistently somewhere.

**Comment by xperimental** [comment]
I have just realized that I (for whatever reason :shrug: ) misinterpreted the date as "the 22nd of October" and not "start of October 2022". Sorry for that, let's see if I can make it work this week :slightly_smiling_face:

**Comment by xperimental** [comment]
I have a "first draft" ready. It's a bit cumbersome to use, mostly depending on "where" your exporter is running. I think I'll be adding the "companion mode" as well, but for now the exporter is "interactive".

The work-in-progress is in the `new-auth` branch (working together with a branch of the same name in my `netatmo-api-go` fork). I have built an image from the current state, which is available as [`ghcr.io/xperimental/netatmo-exporter:new-auth`](https://github.com/xperimental/netatmo-exporter/pkgs/container/netatmo-exporter/43412973?tag=new-auth).

This removes the username and password from the configuration and starts up the exporter "unauthenticated". You need to visit the `/authorize` path on your exporter for the authentication process to start. It should redirect you to the NetAtmo page asking you to authenticate to the exporter application you registered.

There are two new configuration parameters:
- `external-url`, which controls the URL the NetAtmo Interface will redirect you back to after authenticating. This defaults to `http://127.0.0.1:9210` which only works if the exporter is running "locally", so you _probably_ need to change this.
- `token-file`, which, if not empty, makes the exporter save the current token on exit and load it on startup if the file exists.

Theoretically it should be possible to stop/start the exporter with the token saved to the token-file as long as the token is not expired. This is something I want to test myself during the next days.

Please try out the work-in-progress version. Thanks for your feedback.

**Comment by xperimental** [comment]
Today's update is the "authentication companion", a webserver displaying a **very simple** web frontend for authorizing against the NetAtmo API and getting a `token.json` file that can be loaded into `netatmo-exporter`. This is in the [`auth-tool`](https://github.com/xperimental/netatmo-exporter/tree/auth-tool) branch based on yesterday's branch.

You have to compile this yourself (clone, checkout the branch, then do `make auth-tool`) and then run it using:

```console
./auth-tool --client-id $YOUR_CLIENT_ID --client-secret $YOUR_CLIENT_SECRET
```

It will display a local URL to open in your browser which contains additional instructions.

Good luck! :four_leaf_clover:

**Comment by xperimental** [comment]
Apologies for the missing updates, but I am currently dealing with a non-digital emergency. The image mentioned in the earlier comment is running fine for me since posting the comment, so it seems to be fine, just needs a bit of cleanup and documentation.

It's possible that exporters that are running will keep running until the token they use finally expires.

**Comment by alvaroaleman** [comment]
Hey @xperimental thanks for your awesome works on this exporter!

I just wanted to mention that on the app page in the netamo developer portal, there is a "Create Token" button which can be used to create a token. Using the `new-auth` tag and adding the `--token-file` flag pointing to a file whose content is `{"access_token":"$TOKEN"}` worked perfectly fine for me and might avoid the need of implementing the login flow in the exporter.

**Comment by xperimental** [comment]
@alvaroaleman I noticed that button as well, but it only produces an `access_token` with no refresh-token. That way, once the lifetime of the access-token has expired, the exporter has no way of renewing its token automatically.

The tokens produced by the `auth-tool` or the new version of the exporter also contain a refresh-token, so they can be renewed before their original lifetime expires (three hours, if I remember correctly).

Please test with the website-generated token on your end and report what lifetime the token has.

**Comment by alvaroaleman** [comment]
@xperimental oh interesting. Their website doesn't mention about an expiry nor does it show a refresh token. Also, the request to get the actual token that is done by their website gets redirected to another endpoint that returns a bunch of HTML with the token embedded.

I also manually queried the api with curl in the hope that they send any useful response headers back but no luck.

I'll keep it running with this token for now and will report back when/if I get authentication errors. Thanks!

**Comment by alvaroaleman** [comment]
So the website-generated token stopped working after about six hours :upside_down_face: I've generated one including refresh token from the `auth-tool` and that appears to be working fine so far :)

**Comment by j7lb5r** [comment]
Hello there. I found a tutorial how to include my netatmo in a docker container for unraid, but it based on version 1.3.0 :(. After two days of horror with " level=fatal msg="Error in configuration: need a NetAtmo client ID"" I found this thread. I am sorry I am not expert for this, is there any step by step manual how to set up it with new:auth? I generated the Access and refresh token -and now? THX

**Comment by xperimental** [comment]
@j7lb5r Unfortunately I have not yet written the instructions for any of the new login flows. How did you get the access-token/refresh-token? Using the `auth-tool` or using the authentication integrated into the exporter?

The content in this PR is also not yet merged into the master branch or released. I'll try to do this as soon as possible, maybe I have some free cycles during PromCon :wink:

**Comment by j7lb5r** [comment]
I created both tokens at Netatmo Connect Dev Area. I found it below the Client ID generator. There can you generate both tokens.

**Comment by xperimental** [comment]
@j7lb5r Just checked, NetAtmo has updated their web-UI to also provide a refresh-token not only an access-token. That _does_ make things easier (not as good as before, but easier than it was last time I checked). I do not know though, how the OAuth implementation in the exporter handles the missing expiration time. Best case, it immediately uses the refresh-token to get a new token with a known expiry and save that one. Worst case, it just expires like before.

This might be interesting information for you as well, @alvaroaleman :slightly_smiling_face:

**Comment by xperimental** [comment]
@sbrooke Again, apologies for the delay. The release will be coming soon. :crossed_fingers: 

If you're not using Docker, you should be able to build it yourself from the `new-auth` or `auth-tool` branch.

**Comment by sbrooke** [comment]
You have this up above:

_There are two new configuration parameters:

external-url, which controls the URL the NetAtmo Interface will redirect you back to after authenticating. This defaults to http://127.0.0.1:9210 which only works if the exporter is running "locally", so you probably need to change this.
token-file, which, if not empty, makes the exporter save the current token on exit and load it on startup if the file exists._

How do I change the url? I've tried a couple of things and they aren't working.

**Comment by sbrooke** [comment]
Curl from the same system works, just can't access it from somewhere else.

**Comment by xperimental** [comment]
Just released version 1.5.0 to push out all the non-authentication-related changes. Interestingly the old authentication still seems to work fine for me, even though it should have stopped working more than a month ago. Can anyone confirm?

For testing with the new authentication, I have updated the "new-auth" branch with the released changes. The updated Docker image is this one:

```
ghcr.io/xperimental/netatmo-exporter:d59fbff-new-auth
```

**Comment by xperimental** [comment]
@sbrooke You should be able to set the "external-url" using either the `--external-url` command-line flag or the `NETATMO_EXPORTER_EXTERNAL_URL` environment variable. You will need to know the IP-address/hostname the exporter's host is reachable for you (usually something starting with `192.168.` if the exporter is in your local LAN).

**Comment by sbrooke** [comment]
"FATA Error in flags: unknown flag: --external-url" is what I've been getting. I'll install the update and see if it's still happening.

**Comment by sbrooke** [comment]
Ok, got it to launch the Netatmo auth page, authorized it and:

{"error":{"code":404,"message":"method not found"}}

I'll try it again

**Comment by SnorreSelmer** [comment]
Old method still works for me too, no idea how long though…

tir. 6. des. 2022 kl. 01:22 skrev Robert Jacob ***@***.***>:

**Comment by sbrooke** [comment]
Ok, still stuck here. I made the "new-auth" version, pulled the refresh token from Netatmo, set up the token.json file with the refresh token, and now I get a 403 Forbidden. Here's what I'm starting it with. All of the keys have been changed for sharing, of course.

./netatmo-exporter new-auth --client-id abcdefg --client-secret zyxwvu --token-file token.json

The token.json file is using this format:

{"access_token":"123456","expiry":"0001-01-01T00:00:00Z"}

What's interesting is that I didn't add the expiry stuff, the exporter must have. If I use the access token I don't get an error and I finally have stats! Using the refresh token I only get the 403 and no stats. I'll let this run for the day and see if it fails without the refresh token. Any other recommendations, provided it ends up failing on me later today?

**Comment by xperimental** [comment]
@sbrooke The NetAtmo website now issues tokens with both `access_token` and `refresh_token`, if you put that into the `token.json`, it _should work_. Try adding an expiry a few hours in the future if it doesn't work without the expiry.

Two alternatives:
- Try the "auth-tool" from the same-named branch. It will produce a proper `token.json` complete with valid expiry.
- If everything else fails, the old authentication still worked fine last time I checked. It still might.

Need to clean this up and write proper documentation sometime soon, but have other issues to deal with right now.

**Comment by sbrooke** [comment]
I'm waiting to make sure it updates properly, but the problem is that I had "access_token" in the file instead of "refresh_token".

Now, it seems when I run the bash script in systems it's reading the token file for some period of time and then overwriting it with a blank file.

**Comment by sbrooke** [comment]
@xperimental I created a new issue specific to the token.json.

Really appreciate you doing this! Speaking for myself, it's just for my home dashboard, so no need to rush. Thanks!

**Comment by sbrooke** [comment]
That's interesting, because it definitely works the first time with only the refresh token and then updates the file with the rest of the info. I agree that it sounds like some error handling needs to be in place, but it also looks like the refresh isn't happening correctly. I'll run another test today to see if it's still happening right at 3 hours. Maybe the token refresh process isn't happening in the right window for Netatmo to be happy about it?

**Comment by tukane** [comment]
Hi, is it now possible to use the new auth method with the docker image? If yes, with variables do I have to use? I've already generated an Access and Refresh token on the Netatmo API website. Would love to get this working again.

**Comment by SnorreSelmer** [comment]
I use the old method, and I still get data, so confirmed still working (at
least for existing setups)

tir. 13. jun. 2023 kl. 15:09 skrev Robert Jacob ***@***.***>:

**Comment by xperimental** [comment]
@tukane Your error has nothing to do with the discussion in this issue. The error message rather indicates that something's wrong with your local DNS server.

**Comment by sbrooke** [comment]
Started failing with a `Error creating client: oauth2: cannot fetch token: 400\nResponse: {\"error\":\"unauthorized_client\"}` message yesterday. Anyone else or is this just a problem on my end?

**Comment by giuliomagnifico** [comment]
Me too since few hours ago: https://github.com/xperimental/netatmo-exporter/issues/44

Maybe it's only a Netamo issue?

**Comment by xperimental** [comment]
Just checked, I'm getting the same error. :disappointed:

So it seems like Netatmo finally did the change they announced for October last year. I had hoped a bit that they see reason and keep the password authentication for accessing the sensors of the own account. The "client credentials" documentation page also still exists and and still lists October last year as deprecation date.

So, I guess this means I finally need to get the new authentication branch cleaned up, merged and released. I'm sorry for the delay, I should probably have used the past months to do this, but there have been other things that needed my attention.

But there are also good news: the code in the "new-auth" branch still works and the Docker images I produced earlier this year are still available. To save you the scrolling, here's the image name again:

```
ghcr.io/xperimental/netatmo-exporter:d59fbff-new-auth
```

The main downside of the new authentication is that the exporter is not able to start up on its own on the first run or when it has been stopped for a while (and the token expired). Once it is set up, it can run indefinitely without user interaction.

If anyone of you has an idea on how to solve the OAuth authentication in a more "admin-friendly" way, then please let me know. The one improvement that already came up in this thread (but is also not properly documented yet) is using Netatmo's website to get the first token and let the exporter immediately renew it (the token from Netatmo's website does not contain information about the expiry time).

**Comment by giuliomagnifico** [comment]
Okay thanks, but I'm building it from source and I'm also starting the service via systemd  file, so what I have to do to grab the "new auth" branch and what are the new commands to use to launch the exporter? Because the readme is the same in the new branch, or I'm missing something?

At the moment I'm using:

`netatmo-exporter -s xxx -i xxx -p xxx -u giuliomagnifico@gmail.com --refresh-interval 2m`

**Comment by xperimental** [comment]
@giuliomagnifico Updating the documentation is one of the things that is missing and why it's not merged yet. Also there's a few small bugs.

The `--username` and `--password`  parameters do not exist anymore in that version.

You need to pass it the client-id, client-secret and a token-file (so the token can be saved). Once the exporter is running, visit `/authorize` in a browser (for example `http://localhost:9210/authorize`) and it should take you through the OAuth process. Once that is complete the metrics should show up again.

When you exit the exporter, the token will be saved to the `token-file`, so when it is restarted while the token is still valid you do not need to authorize again.

**Comment by giuliomagnifico** [comment]
@xperimental got it thanks but I'm having issues.

When I start the new-auth exporter to get the token I have to specify my Prometheus server url, so I launch it with:

`./netatmo-exporter -s xxx -i xxx --external-url 192.168.1.6`

it opens the Netatmo page correctly

![Screenshot 2023-07-13 at 23 50 33](https://github.com/xperimental/netatmo-exporter/assets/1975176/7a7aca5a-e910-4bcb-be54-b0df83dcb4eb)

but it returns me this error:

`{"error":{"code":404,"message":"method not found"}}`

it should me return the token file? Or what else?

Thanks again and sorry for the questions!

**Comment by xperimental** [comment]
@giuliomagnifico Can you try with setting `--external-url` to a URL, so, in your case, probably something like `http://192.168.1.6:9210`, the callback path will be appended automatically.

The URL just needs to be valid&reachable for _you_, so if you are able to access the exporter as `127.0.0.1:9210`, then you can leave it at the default.

**Comment by xperimental** [comment]
Small update: I have updated the `new-auth` branch with some small fixes and a small UI improvement (I hope :slightly_smiling_face: ). And the Docker image has been updated with the latest changes:

```
ghcr.io/xperimental/netatmo-exporter:0dccd27-new-auth
```

This should also fix #37 and provides some metrics about the token itself, providing some insight into refreshing tokens and validity.

That's it for now :sleeping_bed:

**Comment by giuliomagnifico** [comment]
@xperimental okay found the issue, I have to specify the full url with also the port `--external-url http://192.168.1.6:9210` thanks for the support.

Now the browser says `Authenticated.` **but it doesn't download the token.json** or anything else, in the Netatmo-exporter folder I also don't find the token file. It's my issue?

(I'm using the new exporter updated from your [latest](https://github.com/xperimental/netatmo-exporter/issues/26#issuecomment-1635107423)commits/comment)

**Comment by giuliomagnifico** [comment]
I reply to myself just to say that now I got it working but only by generating manually/via web interface the token, then I created the token.json file and it works

`{"access_token":"1234","refresh_token":"1234"}`

...without the expiration date. I'll try to see what will happend despite the warning `WARN Restored token has no expiry time! Expiry set in 30 Minutes.`

Theorically there's also the refresh token so it should auto-reload it?

> @xperimental okay found the issue, I have to specify the full url with also the port `--external-url http://192.168.1.6:9210` thanks for the support.
> 
> Now the browser says `Authenticated.` **but it doesn't show or download the token.json** or anything else, in the Netatmo-exporter folder I also don't find the token file. It's my issue?
> 
> (I'm using the new exporter updated from your [latest](https://github.com/xperimental/netatmo-exporter/issues/26#issuecomment-1635107423) comment/commits)
> 
> Thanks again for this exporter and your work!

**Comment by SnorreSelmer** [comment]
I’m just chiming in here to say that I’m spun up some services at work
where the official docs said “spin up container X first, run command Y
against it, then spin up the rest of the compose-file.”
If we can have that as the workflow to get this working, then that’s
perfectly fine to me.
I guess we’ll need to attach some local storage (./token) for the token as
well?

fre. 14. jul. 2023 kl. 08:13 skrev giuliomagnifico ***@***.***

**Comment by xperimental** [comment]
@giuliomagnifico 

> Now the browser says Authenticated. but it doesn't show or download the token.json or anything else, in the Netatmo-exporter folder I also don't find the token file. It's my issue?

The `token.json` is currently only touched twice: once on startup to load the token and once on shutdown to write it out again. When there was no `token.json` on startup and you authenticate using the web interface in the exporter the file will not show up until the exporter is stopped.

I made a basic interface in the exporter yesterday at `/`, which shows the current token state and expiry.

> ...without the expiration date. I'll try to see what will happend despite the warning WARN Restored token has no expiry time! Expiry set in 30 Minutes.
>
> Theorically there's also the refresh token so it should auto-reload it?

That's something I added yesterday as well. It will warn about missing bits in the token. When the expiry is missing but there's a refresh token it will set a short expiry (I picked 30min) which will cause it to refresh the token after that time expires. I also found out that it works with _only the refresh token_, it will then refresh to a proper token immediately. Maybe that's the easiest solution for the future.

---

@SnorreSelmer 

> I’m just chiming in here to say that I’m spun up some services at work
where the official docs said “spin up container X first, run command Y
against it, then spin up the rest of the compose-file.”
If we can have that as the workflow to get this working, then that’s
perfectly fine to me.

Thanks for the feedback.

> I guess we’ll need to attach some local storage (./token) for the token as
well?

Yes, if you want to persist the token across restarts it now needs some kind of storage. I made my exporter a StateFulSet with one replica.

---

btw: one quirk in the exporter that is because on how it's caching the NetAtmo data internally is that after authentication is successful, the metrics will only show up on **the second scrape**. This has been the case with the previous password authentication as well, but might catch people off-guard when now looking at "does the authentication work properly".

**Comment by tschmitt1605** [comment]
Hi there! I have the same problems as all of you, but I continue to get errors after your solutions. 
I use the netatmo-exporter with Docker and I'm not quite sure how to include the external-url in the ENVs. If I simply create external-url there and specify my IP I am still redirected to 127.0.0.1. 
I have also tried to create the token.json manually, but the file is not even considered by the exporter.

BTW: Thanks for your great work @xperimental !

**Comment by tschmitt1605** [comment]
I got it working! I simply did not use ENV variables now, but used CMD variables.

My only problem is that I now have to re-authenticate every time I reboot (which fortunately is not often). No data is stored in token.json for me. Is there any way to solve this?

**Comment by xperimental** [comment]
@tschmitt1605 

> I got it working! I simply did not use ENV variables now, but used CMD variables.

The configuration options can also be set with environment variables. They do look a bit different than the command-line arguments, though. Both are listed [here](https://github.com/xperimental/netatmo-exporter/blob/0dccd27aca08d22e76d3bb0b170f77cf04361fbc/internal/config/config.go#L15-L33).

> No data is stored in token.json for me. Is there any way to solve this?

You mention Docker. Do write the file to a location which is a volume? Otherwise it will not persist across recreations of the container. It should still persist across restarts of the container though.

**Comment by sbrooke** [comment]
Pulled the new-auth branch again and created the simple token.json referenced by guiliomagnifico. Restarted the service and it's all good for now. I'll let you know if it quits after a period.

Thanks for jumping on this quickly! My wife also thanks you for getting her weather stats back! :)

**Comment by giuliomagnifico** [comment]
@xperimental @sbrooke I can confirm that manually adding the `access` and `refresh` tokens to the `token.json` is working (after 12h) and after a reboot.

**Comment by tschmitt1605** [comment]
@xperimental 


> The configuration options can also be set with environment variables. They do look a bit different than the command-line arguments, though. Both are listed [here](https://github.com/xperimental/netatmo-exporter/blob/0dccd27aca08d22e76d3bb0b170f77cf04361fbc/internal/config/config.go#L15-L33).

Ok, thanks for the link.

> You mention Docker. Do write the file to a location which is a volume? Otherwise it will not persist across recreations of the container. It should still persist across restarts of the container though.

I have a bind from /docker/netatmo-exporter/token.json to /token.json . So it should work normally.



Meanwhile, I have another error that occurred overnight. I get the following error: Error getting token: Post "https://api.netatmo.net/oauth2/token": context canceled.



I hope you can help me.

**Comment by xperimental** [comment]
@sbrooke @giuliomagnifico Reading this thread I come to the conclusion that the easiest route for adding the token is probably getting it from NetAtmo's website and entering it into the exporter's configuration manually, so I'll try to make that route simpler (the `token.json` still feels a bit cumbersome, even if it works).

So far, I had no issues with using `/authorize` of the exporter to go through the OAuth process and I wonder if I'm missing something that breaks it for you, because I know how it _should work_. Would you like to try out the web-flow again to either confirm that it works or tell me where it breaks for you, now that you have a working configuration?

---

@tschmitt1605 

> Meanwhile, I have another error that occurred overnight. I get the following error: Error getting token: Post "https://api.netatmo.net/oauth2/token": context canceled.

This sounds like a timeout. Did it only happen once? Did the exporter recover from it?

---

In case you missed it: the `new_auth` branch also adds new metrics for checking on the token:

- `netatmo_exporter_token_valid`, which is 1 when the exporter has a valid token
- `netatmo_exporter_token_expiry_time`, which contains the timestamp when the token will expire. `netatmo_exporter_token_expiry_time - time()` should yield a sawtooth curve with a peak at ~3h (the tokens seem to currently have a lifetime of 3h), good for monitoring the refresh of the token

If everything goes as planned, I should have the `2.0.0` out tomorrow. I probably don't have time for coding today.

**Comment by giuliomagnifico** [comment]
@xperimental 
> Would you like to try out the web-flow again to either confirm that it works or tell me where it breaks for you, now that you have a working configuration?

For me the process of OAuth works but it doens't download the token file or nothing else, I see only an `Authorized` word after I press "Yes I accept" on the Netamo website. This is why I went to manually ad the data to the `token.json` file.

**Comment by tschmitt1605** [comment]
> This sounds like a timeout. Did it only happen once? Did the exporter recover from it?

It happened only once, but I can't give you more data about it, because I restarted the container immediately and included the new ENV varibals. Now retrieving the data from token.json works as well. 



> For me the process of OAuth works but it doens't download the token file or nothing else, I see only an `Authorized` word after I press "Yes I accept" on the Netamo website. This is why I went to manually ad the data to the `token.json` file.

I have the same problem and I have also entered the dates manually.




> If everything goes as planned, I should have the `2.0.0` out tomorrow. I probably don't have time for coding today.

Nice! Thank you!

**Comment by sbrooke** [comment]
Sad to report it zeroed out the token.json file on me only a little while after it started working. When I manually create the token.json the process has a message about "no expiry, setting it to 30 minutes".

I just tried the OAuth and it worked, and updated the token.json file. It also added an expiry date to it. However, when I was going through the process it tried to do a callback to 127.0.0.1 and failed. I'm running this on a Ubuntu VM and did the OAuth from my laptop. However, it's not actually pulling stats. I'm getting the following:

`Jul 15 12:32:45 netatmo systemd[1]: Starting Netatmo...
Jul 15 12:32:45 netatmo bash[994]: level=info msg="Listen on :9210..."
Jul 15 12:32:53 netatmo bash[994]: level=error msg="Error during refresh: bad HTTP return code: 403"`

Close to 30 minutes and it zeroed the file again.

**Comment by xperimental** [comment]
@sbrooke 

> Close to 30 minutes and it zeroed the file again.

I'm confused by this. The exporter only touches the file _on startup and shutdown_, not while it is running. Is something else (not the exporter itself) causing the exporter to restart or overwriting the file?

**Comment by xperimental** [comment]
@giuliomagnifico 

> For me the process of OAuth works but it doens't download the token file or nothing else, I see only an Authorized word after I press "Yes I accept" on the Netamo website.

The token file is not written until the exporter exits. If you see the callback (the page with "Authorized" on it) then the file _will not immediately appear_. The exporter should have the token in memory though and be able to scrape metrics.

**Comment by sbrooke** [comment]
> @sbrooke
> 
> > Close to 30 minutes and it zeroed the file again.
> 
> I'm confused by this. The exporter only touches the file _on startup and shutdown_, not while it is running. Is something else (not the exporter itself) causing the exporter to restart or overwriting the file?

I'm running it as a service. I'm not using Docker. Don't know if that has anything to do with it. Also, I don't know if it's missing a step when I try to do the authentication through the exporter's web interface. In other words, I click the link you set up in the root web page, it takes me to Netatmo to auth, I complete that and it returns me to 127.0.0.1, which is wrong because I'm doing this from my PC and not from the Exporter server. I get an error message there, but it does update the token.json to include the updated info.

**Comment by sbrooke** [comment]
I can also see at the root web page that:

`You have a token.

Token is valid until 2023-07-16 14:55:30.281143226 +0000 UTC (27m15.683161881s)`

**Comment by sbrooke** [comment]
Sorry for the multiple posts. I just copied the old (non-expiry) token.json in and rebooted and I have the above message, but it's not connecting to Netatmo successfully. The Exporter has a stats page but none of the Netatmo stats are listed and the service is showing a 403 error message.

I'm going to try removing the token.json and going through the web page auth process again.

**Comment by sbrooke** [comment]
It's quite possible my service is formatted incorrectly:

_/home/netatmo/netatmo-exporter/netatmo-exporter --client-id BLAH --client-secret BLEH --token-file /home/netatmo/netatmo-exporter/token.json_

**Comment by xperimental** [comment]
@sbrooke 

> In other words, I click the link you set up in the root web page, it takes me to Netatmo to auth, I complete that and it returns me to 127.0.0.1, which is wrong because I'm doing this from my PC and not from the Exporter server.

If you're not accessing the `/authorize` URL from `127.0.0.1` then you need to update the `--external-url` in the exporter, so that it generates the correct URL for NetAtmo to redirect you to. The "flow" is something like this:

- you access the exporter at `/authorize`, it will redirect you to the NetAtmo site
- you click "Yes, I accept" on NetAtmo's website
- it will try to redirect you back to the exporter (to a `/callback` URL). This is the important step, as this will enable the exporter to get the token. If it was successful, you should see a "Authorized" message
- the exporter now has the token and should be able to scrape the metrics (it will **not** write the token-file at this point)
- when the exporter shuts down, it will write its _current_ token to the token-file

> I just copied the old (non-expiry) token.json in and rebooted and I have the above message, but it's not connecting to Netatmo successfully.

The tokens issues by NetAtmo are only valid for ~3h, this is also true for the ones generated using their website. This means that you can not "reuse" an old token, even if you created it manually using the website. When the token has no expiry, the exporter will "guess" a lifetime of 30min, to renew the token earlier. When the exporter has a token with _a refresh-token_, it will automatically renew the token when it expires.

**Comment by xperimental** [comment]
@sbrooke Two questions that just occurred to me:

1) What "scope" do you select when creating the token on the website? (`read_station` is the one the exporter currently needs)
2) What information do you put in the `token.json` file when you create it manually? Just the access-token or also the refresh-token?

---

> I'm running it as a service. I'm not using Docker. Don't know if that has anything to do with it.

No, it should also work without using Docker. In fact that's how I'm testing locally as well. It should make it _simpler_ to configure instead of harder, as you don't need to worry about the container namespacing.

---

I'm currently adding a way to "simply" get the token manually created on the website into the exporter.

**Comment by sbrooke** [comment]
Ok, ran it directly with --external-url http://192.168.1.19:9210 and the authentication succeeded this time. And I think it took a restart but I'm getting weather stats now. Token shows ~3 hours. Running the service without the --external-url appears to be working, although it's not certain yet.

Crossing my fingers...I'll let you know in a couple of hours.

**Comment by xperimental** [comment]
> Running the service without the --external-url appears to be working, although it's not certain yet.

The `external-url` is only needed for the last part of the authentication (the callback), so once the exporter has a refreshable token you should be fine.

**Comment by xperimental** [comment]
A few new changes have been incorporated into `new_auth` and I have built a new Docker image to test with:

```
ghcr.io/xperimental/netatmo-exporter:db281b5-new-auth
```

The main change since the last build is that I have extended the "homepage". There's a textbox for filling in the `refresh_token` from the NetAtmo developer console now, which _should_ make setup easier. I'd be thankful if someone would be able to test this and give me some feedback (please don't provide any feedback on how the page _looks_, I know it is awful :wink: ).

I have also extended the info and warning messages for loading/saving the token file a bit. The Docker image has been extended to provide a volume for the token-file by default. This can be customized, of course, but might make it easier to setup.

---

If the new version works for most of you and there are no showstoppers anymore then I think this might be the version that will get the initial 2.0 release. :slightly_smiling_face:

**Comment by xperimental** [comment]
There's also [a new document](https://github.com/xperimental/netatmo-exporter/blob/new-auth/doc/authentication.md)  explaining how to authenticate. Feedback appreciated. :slightly_smiling_face:

**Comment by theonlydoo** [comment]
> There's also [a new document](https://github.com/xperimental/netatmo-exporter/blob/master/doc/authentication.md) explaining how to authenticate. Feedback appreciated. 🙂

Tested and validated! works perfectly, thanks!

**Comment by tschmitt1605** [comment]
Everything works for me as well. I think you can publish the current new auth as main now. 

Thanks for your commitment and the exporter.

**Comment by sbrooke** [comment]
Grabbed the latest main this morning and it's working fine. If I run it manually. Doesn't like to run as a service. Probably a config problem on my end, so I'll sort that out.

**Comment by tschmitt1605** [comment]
> Grabbed the latest main this morning and it's working fine. If I run it manually. Doesn't like to run as a service. Probably a config problem on my end, so I'll sort that out.

I will try it later if I have the same problems when I run it as a service?


@xperimental 

Do you need help with the docs?

**Comment by xperimental** [comment]
@sbrooke 

> Doesn't like to run as a service.

Do you get any error? Based on your previous messages, I'm assuming systemd service unit... do you use a user that has read/write access to the path you use as a `token-file`? Any of the sandboxing features of systemd active?

@tschmitt1605 

> Do you need help with the docs?

My hope is that the updated readme and the new files in `doc/` are sufficient. :slightly_smiling_face: 

> I think you can publish the current new auth as main now.

I have already merged the PR on Sunday. The 2.0.0 release _might_ happen today.

---

One thought I had while working on this issue is whether it would make sense to add some "template configuration files" in a `contrib/` folder... I'm using a `StatefulSet`, so I can provide this. There seem to be a few people using systemd to run it as a normal service (which I approve of as well :+1: :wink: ) ...

There has also always been the question of "should I build binaries as well in addition to the images"... I think that might be something to improve post-2.0.0 as well.

**Comment by sbrooke** [comment]
> Do you get any error? Based on your previous messages, I'm assuming systemd service unit... do you use a user that has read/write access to the path you use as a token-file? Any of the sandboxing features of systemd active?

I'll go back and look at the permissions. What's odd is that it's touching the token.json file, but then blanking it. In other words, the token is present in the file until something causes it to be overwritten with an empty file. Doesn't seem to be an issue with the ability to write the file, just that it doesn't have a token to write into it.

Here's the service:

`[Unit]
After=network.target
Description=Netatmo

[Service]
Type=forking
User=netatmo
WorkingDirectory=/usr/bin
ExecStart=/usr/bin/bash -c "/usr/bin/netatmo.sh"
Restart=always

[Install]
WantedBy=multi-user.target`

Probably something wrong in what I have there.

**Comment by sbrooke** [comment]
The netatmo.sh file is simply running the command string I used to launch it. Perhaps the service isn't seeing it as a running program and restarting it constantly? New instance wipes out the token.json to start over again?

**Comment by xperimental** [comment]
@sbrooke If you like, we can debug your case, but I think it would be nice to have a separate issue for it. Would you like to create one?

Why do you need the shell script and `Type=forking`? Only thing I can think of right now is setting up the environment and I'd think using `Environment` or `EnvironmentFile` directives is better suited to that.

And when you say the token-file is "blanked" is it really completely empty (0b) or is it just a JSON with empty values?

**Comment by xperimental** [comment]
`netatmo-exporter 2.0.0` has just been released :rocket:
Thank you all for your patience!

**Comment by sbrooke** [comment]
> @sbrooke If you like, we can debug your case, but I think it would be nice to have a separate issue for it. Would you like to create one?
> 
> Why do you need the shell script and `Type=forking`? Only thing I can think of right now is setting up the environment and I'd think using `Environment` or `EnvironmentFile` directives is better suited to that.
> 
> And when you say the token-file is "blanked" is it really completely empty (0b) or is it just a JSON with empty values?

Totally agree, this issue is closed. Thank you!

**Comment by m-arx** [comment]
I cannot get it running as a docker container. I have build an image with 2.0.0

I have created an env file with this content:

`NETATMO_CLIENT_ID='lesecretsauce'
NETATMO_CLIENT_PASSWORD='cannotreadme'
NETATMO_CLIENT_SECRET='gibberish'
NETATMO_CLIENT_USERNAME='me@world.com'
NETATMO_EXPORTER_EXTERNAL_URL='http://192.168.2.40:9210'`


I have tried the same without the external url, tunnelling the 9210 port from my docker machine to my notebook and calling it from 127.0.0.1, I always end up with "Error processing code: user did not accept" after clicking on "authorize".

also I have tried creating a token.json with a start token from the development.netatmo.com site (and mounting it into the docker container and creating an env for it).

**Comment by tschmitt1605** [comment]
@m-arx 
You already have too many ENV Varibals. With the new authentication, username and password are no longer needed. 
The following ENVs are sufficient:
`NETATMO_CLIENT_ID='lesecretsauce' NETATMO_CLIENT_SECRET='gibberish' NETATMO_EXPORTER_EXTERNAL_URL='http://192.168.2.40:9210'`

If you want to save your token.json file separately you also have to add the file path: `NETATMO_EXPORTER_TOKEN_FILE=/token.json`
But then you have to include the token.json from the system as well.



If you have any questions or it still doesn't work, please contact me again.

**Comment by m-arx** [comment]
> If you have any questions or it still doesn't work, please contact me again.

Thank you, I have now changed my env file according to the above, but when clicking I get a 404 from netatmo when calling the url "https://api.netatmo.net/oauth2/'http://192.168.2.40:9210'/auth/callback?error=invalid_client&state=definitelyrandom"

The return in the browser is "{"error":{"code":404,"message":"method not found"}}"

**Comment by tschmitt1605** [comment]
> The return in the browser is "{"error":{"code":404,"message":"method not found"}}"

If you get this error, you may have set the wrong external-url. I had taken the one from your example. Check the iP again and then everything should work.

**Comment by xperimental** [comment]
@m-arx Didn't see it yesterday, but this has happened previously in another question: The values specified in the environment file are forwarded as-is to the exporter. Do not put quotes (`'` or `"`) around the values.

If this needs further debugging, can you please create a new issue to discuss this in?

**Comment by giuliomagnifico** [comment]
Sorry for the delay @xperimental, I had the time to test the new branch v2.0 compiled - not docker and not the experimental one with the token- only now (that I'm not at work), and I can confirm that it works fine always using/filling manually the `token.json` file.

What I want to suggest is to have the execute permission on the `token.json` file, like a 755, because without it I'm not sure that the exporter can read it (I had some issues).

```
root@Grafana:~/netatmo-exporter# service netatmo-exporter status
● netatmo-exporter.service - netatmo-exporter
     Loaded: loaded (/etc/systemd/system/netatmo-exporter.service; enabled; vendor preset: enabled)
     Active: active (running) since Sat 2023-07-22 09:29:56 CEST; 34s ago
   Main PID: 127267 (netatmo-exporte)
      Tasks: 9 (limit: 4531)
        CPU: 161ms
     CGroup: /system.slice/netatmo-exporter.service
             └─127267 /root/netatmo-exporter/netatmo-exporter -s 123 -i 123 -
-token-file token.json

Jul 22 09:29:56 Grafana systemd[1]: Started netatmo-exporter.
Jul 22 09:29:56 Grafana netatmo-exporter[127267]: level=info msg="Loaded token from token.json."
Jul 22 09:29:56 Grafana netatmo-exporter[127267]: level=info msg="Listen on :9210..."
```

**Comment by xperimental** [comment]
@giuliomagnifico Good to hear that it works for you.

> What I want to suggest is to have the execute permission on the token.json file

While the exporter needs both read and write access to the token file, it does not need the execute bit to be set. If your current configuration works with the file having a mode of 0755 (`-rwxr-xr-x`) it should work just the same with 0644 (`-rw-r--r--`). I don't know what your service unit file looks like, but I would recommend running the exporter as a non-root user and putting the files outside of `/root`. I have added [an example unit file](https://github.com/xperimental/netatmo-exporter/commit/d0dbf0d98040eb504ff3bada67c9667b03d7cbd0#diff-f99b7c545fd30ba82d1ac3aab7c5825bee58a0c6de7c4cd9b7d4209d4288ccd6) a few days ago, which you can use as a starting-point.

We can talk more about this, but please open a separate issue for that. :slightly_smiling_face:

**Comment by SnorreSelmer** [comment]
Finally got my server back up after moving house, updated the container to the latest version, banged my head against the wall because I didn't read the instructions properly (NETATMO_EXPORTER_EXTERNAL_URL needs both "http://" and port-number, duh), but the setup works great and I have data flowing in just fine! Two thumbs up!
